src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -740,220 +731,438 @@ public void visitSwitchExpression(JCSwitchExpression tree) {
-                        UniqueType type = new UniqueType(dpat.type, types);
-                        List<JCRecordPattern> augmentedPatterns =
-                                deconstructionPatternsByType.getOrDefault(type, List.nil())
-                                                                 .prepend(dpat);
-
-                        deconstructionPatternsByType.put(type, augmentedPatterns);
+                    return new RecordPattern(record.type, componentTypes, nestedDescriptions);
+                } else {
+                    throw Assert.error();
+                }
+            }
+        }
+
+        record BindingPattern(Type type) implements PatternDescription {
+            @Override
+            public int hashCode() {
+                return type.tsym.hashCode();
+            }
+        }
+
+        record RecordPattern(Type recordType, int _hashCode, Type[] fullComponentTypes, PatternDescription... nested) implements PatternDescription {
+
+            public RecordPattern(Type recordType, Type[] fullComponentTypes, PatternDescription[] nested) {
+                this(recordType, hashCode(-1, recordType, fullComponentTypes, nested), fullComponentTypes, nested);
+            }
+
+            public int hashCode() {
+                return _hashCode;
+            }
+
+            public int hashCode(int excludeComponent) {
+                return hashCode(excludeComponent, recordType, fullComponentTypes, nested);
+            }
+
+            public static int hashCode(int excludeComponent, Type recordType, Type[] fullComponentTypes, PatternDescription... nested) {
+                int hash = 5;
+                hash =  41 * hash + recordType.tsym.hashCode();
+                for (int  i = 0; i < nested.length; i++) {
+                    if (i != excludeComponent) {
+                        hash = 41 * hash + nested[i].hashCode();
                     }
+                }
+                return hash;
+            }
+        }
+
+        private boolean exhausts(JCExpression selector, List<JCCase> cases) {
+            List<PatternDescription> patterns = List.nil();
+            Map<Symbol, Set<Symbol>> enum2Constants = new HashMap<>();
+            for (JCCase c : cases) {
+                if (!TreeInfo.unguardedCase(c))
+                    continue;
 
-                    default -> {
-                        Assert.check(labelValue instanceof JCExpression, labelValue.getTag().name());
-                        JCExpression expr = (JCExpression) labelValue;
-                        if (expr.hasTag(IDENT) && ((JCIdent) expr).sym.isEnum())
-                            coveredSymbols.add(((JCIdent) expr).sym);
+                for (var l : c.labels) {
+                    if (l instanceof JCPatternCaseLabel patternLabel) {
+                        for (Type component : components(selector.type)) {
+                            patterns = patterns.prepend(PatternDescription.from(types, component, patternLabel.pat));
```

Comment:
I noted that this code ends up adding redundant pattern descriptions to the list - for instance:

```java
class Test {
    sealed interface I1 permits B, C { }
    sealed interface I2 permits B, C { }

    static final class B implements I1, I2 { }
    static final class C implements I1, I2 { }

    <Z extends I1 & I2> int test(Z z) {
        return switch (z) {
            case B c -> 2;
            case C d -> 3;
        };
    }
}
```

In this case the list ends up with 6 elements, [ B, B, B, C, C, C ]. Given that the complexity of the algorithm depends on the number of patterns in the list, it would probably be better to use a set here and try to make the list as small as possible from early on.

Reply:
I've also found an infinite loop with this:

```java
class Test {
    sealed interface I0 permits I1, I2 { }
    sealed interface I00 permits I1, I2 { }

    sealed interface I1 extends I0, I00 permits B, C { }
    sealed interface I2 extends I0, I00 permits B, C { }

    static final class B implements I1, I2 { }
    static final class C implements I1, I2 { }

    int test(Object o) {
        return switch (o) {
            case B c -> 2;
            case C d -> 3;
        };
    }
}
```

Reply:
> I've also found an infinite loop with this:

I believe this has to do with the fact that the list of pattern is not a set - so we can end up adding the same pattern descriptions over and over.

Reply:
Thanks for the testcase!

I've tried to fix this by (avoiding to add redundant binding patterns):
https://github.com/openjdk/jdk/pull/13074/commits/7e6ed619d89784c49a96984c0f74176a9e0bcf63

Sets might be a good move eventually, but so far it seemed to hide problems, which quite probably needed to be solved anyway, so I kept the list impl for now.